javatests/arcs/core/storage/StorageProxyTest.kt

```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Test
+    fun propagatesStorageFullModelToReaders() = runBlockingTest {
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, mockCrdtModel)
+        val readHandle = handle("testReader", storageProxy, true)
+        mockCrdtModel.appliesOpAs(mockCrdtOperation, true)
+
+        storageProxy.registerHandle(readHandle)
+        storageProxy.onMessage(ProxyMessage.ModelUpdate(mockCrdtData, null))
+
+        verify(readHandle.callback!!).onSync()
+    }
+
+    @Test
+    fun propagatesStorageSyncReqToStorage() = runBlockingTest {
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, mockCrdtModel)
+        mockCrdtModel.appliesOpAs(mockCrdtOperation, true)
+
+        storageProxy.onMessage(ProxyMessage.SyncRequest(null))
+
+        val modelUpdate = ProxyMessage.ModelUpdate<CrdtData, CrdtOperation, String>(mockCrdtData, null)
+        verify(mockStorageEndpoint).onProxyMessage(modelUpdate)
+    }
+
+    @Test
+    fun propagatesUpdatesToReadersAndNotToWriters() = runBlockingTest {
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, mockCrdtModel)
+        val readHandle = handle("testReader", storageProxy, true)
+        val writeHandle = handle("testWriter", storageProxy, false)
+        mockCrdtModel.appliesOpAs(mockCrdtOperation, true)
+
+        storageProxy.registerHandle(readHandle)
+        storageProxy.registerHandle(writeHandle)
+        assertThat(storageProxy.applyOp(mockCrdtOperation)).isTrue()
+
+        verify(readHandle.callback!!).onUpdate(mockCrdtOperation)
+        verifyNoMoreInteractions(writeHandle.callback!!)
+    }
+
+    @Test
+    fun failedApplyOpTriggersSync() = runBlockingTest {
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, mockCrdtModel)
+        val readHandle = handle("testReader", storageProxy, true)
+        mockCrdtModel.appliesOpAs(mockCrdtOperation, false)
+
+        storageProxy.registerHandle(readHandle)
+        assertThat(storageProxy.applyOp(mockCrdtOperation)).isFalse()
+
+        val syncReq = ProxyMessage.SyncRequest<CrdtData, CrdtOperation, String>(null)
+        verify(mockStorageEndpoint).onProxyMessage(syncReq)
+    }
+
+    @Test
+    fun getParticleViewReturnsSyncedState() = runBlockingTest {
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, mockCrdtModel)
+        Mockito.`when`(mockCrdtModel.consumerView).thenReturn("someData")
+        val readHandle = handle("testReader", storageProxy, true)
+        mockCrdtModel.appliesOpAs(mockCrdtOperation, true)
+
+        storageProxy.registerHandle(readHandle)
+        assertThat(storageProxy.onMessage(ProxyMessage.Operations(listOf(mockCrdtOperation), null)))
+            .isTrue()
+        val view = storageProxy.getParticleView()
+
+        assertThat(view.value).isEqualTo("someData")
+        assertThat(view.versionMap).isEqualTo(VersionMap())
+    }
+
+    private fun handle(name: String,
```

Comment:
`newHandle` or `createHandle`

Reply:
Done. My 2.5 years of daily Go programming are leaking..